                                   CrossBorder SQL project
    
    Case-Study:
        
        
CrossBorder Films is one of the leading DVD rental companies that offers film rental services across different countries. Having completed the 10Alytics SQL training module, the head of
data and Analytics of CrossBorder Films has hired you as an analyst to help them make sense of their business data. They have sent you their dvdrental database(Same dvdrental database you
have been working with) to work with.

They require you to answer the following questions about their film rental business
performance:
    
Question 1: Who are the top loyal customers of CrossBorder Films and where do they live?
    
Question 2. Generate a report of film rental performance over the years looking at how many
films were returned late, early and on time?

Question 3: What are the top 3 highest in-demand genres of film over the years and what are
their total sales?

Question 4: Identify the top 5 countries where CrossBorder Films had the highest volume of
sales along-side the countries where they had the largest customer base.

Question 5: Identify the top 5 countries where CrossBorder Films had the lowest customer
base.

Question 6:For each of the questions above, explain in one paragraph what insight can be concluded from
the result of your query.

Load Ipython SQL

In [1]:
%load_ext sql

Connect to the dvdrental database in postgress

In [2]:
# We're connecting to the PostgreSQL database using the PostgreSQL connection string
# The PostgreSQL string is in this format: %sql dialect+driver://username:password@host:port/databaseName

%sql postgresql://chris:admin1234@localhost/dvdrental

In [42]:
%%sql
SELECT
    c.customer_id,
    CONCAT(c.first_name, ' ', c.last_name) AS customer_name,
    a.address,
    ci.city,
    co.country
FROM
    customer c
JOIN
    address a ON c.address_id = a.address_id
JOIN
    city ci ON a.city_id = ci.city_id
JOIN
    country co ON ci.country_id = co.country_id
JOIN
    film f ON f.film_id = f.film_id
ORDER BY
    f.rental_rate DESC
LIMIT 10;


 * postgresql://chris:***@localhost/dvdrental
10 rows affected.


[(524, 'Jared Ely', '1003 Qinhuangdao Street', 'Purwakarta', 'Indonesia'),
 (524, 'Jared Ely', '1003 Qinhuangdao Street', 'Purwakarta', 'Indonesia'),
 (524, 'Jared Ely', '1003 Qinhuangdao Street', 'Purwakarta', 'Indonesia'),
 (524, 'Jared Ely', '1003 Qinhuangdao Street', 'Purwakarta', 'Indonesia'),
 (524, 'Jared Ely', '1003 Qinhuangdao Street', 'Purwakarta', 'Indonesia'),
 (524, 'Jared Ely', '1003 Qinhuangdao Street', 'Purwakarta', 'Indonesia'),
 (524, 'Jared Ely', '1003 Qinhuangdao Street', 'Purwakarta', 'Indonesia'),
 (524, 'Jared Ely', '1003 Qinhuangdao Street', 'Purwakarta', 'Indonesia'),
 (524, 'Jared Ely', '1003 Qinhuangdao Street', 'Purwakarta', 'Indonesia'),
 (524, 'Jared Ely', '1003 Qinhuangdao Street', 'Purwakarta', 'Indonesia')]

In [30]:
Question 2. Generate a report of film rental performance over the years looking at how many films were returned late, early and on time?
%%sql 
SELECT 
    DATE_PART('year', r.rental_date) AS rental_year,
    COUNT(CASE WHEN r.return_date > r.rental_date + INTERVAL '2 days' THEN 1 END) AS late_returns,
    COUNT(CASE WHEN r.return_date < r.rental_date + INTERVAL '2 days' THEN 1 END) AS early_returns,
    COUNT(CASE WHEN r.return_date = r.rental_date + INTERVAL '2 days' THEN 1 END) AS on_time_returns
FROM 
    rental r
GROUP BY 
    rental_year
ORDER BY 
    rental_year;

 * postgresql://chris:***@localhost/dvdrental
3 rows affected.


[(2005.0, 13184, 2675, 1), (2006.0, 0, 0, 0), (2021.0, 2, 0, 0)]

In [23]:
# Question 3: What are the top 3 highest in-demand genres of film over the years and what are their total sales?

%%sql
SELECT 
    c.name AS genre,
    SUM(p.amount) AS total_sales
FROM 
    film f
JOIN 
    film_category fc ON f.film_id = fc.film_id
JOIN 
    category c ON fc.category_id = c.category_id
JOIN 
       inventory i ON f.film_id = i.film_id
JOIN 
    rental r ON i.inventory_id = r.inventory_id
JOIN 
    payment p ON r.rental_id = p.rental_id
GROUP BY 
    genre
ORDER BY 
    total_sales DESC
LIMIT 3;

 * postgresql://chris:***@localhost/dvdrental
3 rows affected.


[('Sports', Decimal('4892.19')),
 ('Sci-Fi', Decimal('4336.01')),
 ('Animation', Decimal('4245.31'))]

In [28]:
# Question 4: Identify the top 5 countries where CrossBorder Films had the highest volume of sales along-side the countries where they had the largest customer base.
%%sql
WITH sales_by_country AS (
    SELECT 
        co.country_id, 
        co.country,
        SUM(p.amount) AS total_sales
    FROM 
        payment p
    JOIN 
        customer c ON p.customer_id = c.customer_id
	JOIN 
        address a ON c.address_id = a.address_id
    JOIN 
        city ci ON a.city_id = ci.city_id
    JOIN 
        country co ON ci.country_id = co.country_id
    GROUP BY 
        co.country_id, co.country
    ORDER BY 
        total_sales DESC
	LIMIT 5
),
customers_by_country AS (
    SELECT 
        co.country_id, 
        co.country,
        COUNT(c.customer_id) AS customer_count
    FROM 
        customer c
    JOIN 
        address a ON c.address_id = a.address_id
    JOIN 
        city ci ON a.city_id = ci.city_id
    JOIN 
	    country co ON ci.country_id = co.country_id
    GROUP BY 
        co.country_id, co.country
    ORDER BY 
        customer_count DESC
    LIMIT 5
)
SELECT 
    s.country AS top_sales_country,
    s.total_sales AS total_sales_in_country,
    c.country AS top_customers_country,
    c.customer_count AS total_customers_in_country
FROM 
    sales_by_country s
JOIN 
    customers_by_country c ON s.country_id = c.country_id;


 * postgresql://chris:***@localhost/dvdrental
5 rows affected.


[('India', Decimal('6034.78'), 'India', 60),
 ('China', Decimal('5251.03'), 'China', 53),
 ('United States', Decimal('3685.31'), 'United States', 37),
 ('Japan', Decimal('3122.51'), 'Japan', 33),
 ('Mexico', Decimal('2984.82'), 'Mexico', 30)]

In [27]:
# Question 5: Identify the top 5 countries where CrossBorder Films had the lowest customer base.
%%sql
WITH customers_by_country AS (
    SELECT 
        co.country_id, 
        co.country,
        COUNT(c.customer_id) AS customer_count
    FROM 
        customer c
    JOIN 
        address a ON c.address_id = a.address_id
    JOIN 
        city ci ON a.city_id = ci.city_id
    JOIN 
        country co ON ci.country_id = co.country_id
    GROUP BY 
        co.country_id, co.country
	ORDER BY 
        customer_count ASC
    LIMIT 5
)
SELECT 
    country,
    customer_count
FROM 
    customers_by_country;

 * postgresql://chris:***@localhost/dvdrental
5 rows affected.


[('Afghanistan', 1),
 ('Finland', 1),
 ('Virgin Islands, U.S.', 1),
 ('Anguilla', 1),
 ('Liechtenstein', 1)]

Question 6:
For each of the questions above, explain in one paragraph what insight can be concluded from the result of your query.

2006 no early return, no late_return and no. on_time return
2005 late returns 13184, early return 2675  and 1 on_time return
2021 2 late_return, 0 early and 0 on_time return
Highest in-demand genres of films are sports, Sci—Fi and Animation
Top sales was in India and last top 5 is Mexico.
Top 5 countries with lowest customer base are Afghanistan, Finland, Virgin Island, U.S, Anguilla and Leichesten.
